src/dfx/src/lib/models/canister.rs

```diff
@@ -102,13 +103,50 @@ impl Canister {
         self.builder.generate(pool, &self.info, build_config)
     }
 
-    #[context("Failed while trying to apply metadata for canister '{}'.", self.info.get_name())]
-    pub(crate) fn apply_metadata(&self, logger: &Logger) -> DfxResult {
-        let mut metadata_sections = self.info.metadata().sections.clone();
-        // Default to write public candid:service unless overwritten
-        if (self.info.is_rust() || self.info.is_motoko())
-            && !metadata_sections.contains_key(CANDID_SERVICE)
+    #[context("Failed to post-process wasm of canister '{}'.", self.info.get_name())]
+    pub(crate) fn wasm_post_process(
+        &self,
+        logger: &Logger,
+        build_output: &BuildOutput,
+    ) -> DfxResult {
+        let build_output_wasm_path = match &build_output.wasm {
+            WasmBuildOutput::File(p) => p,
+            WasmBuildOutput::None => {
+                return Ok(());
+            }
+        };
+        let wasm_path = self.info.get_build_wasm_path();
+        dfx_core::fs::composite::ensure_parent_dir_exists(&wasm_path)?;
+        let info = &self.info;
+        if info.is_pull() || info.is_remote() {
+            // We won't install such canisters with `dfx canister install` or `dfx deploy`.
+            // So no need to copy the wasm.
+            // Pull type canisters have empty `build_output_wasm_path`. Impossible to copy it.
+            return Ok(());
+        }
```

Comment:
Do you want to consolidate this check with the one above?  IIUC, `info.is_pull()` can never be true here, because build_output.wasm will always be WasmBuildOutput::None.  And `Pull type canisters have empty build_output_wasm_path` is no longer accurate.